app/controllers/reports_controller.rb

```diff
@@ -5,20 +5,69 @@ def dashboard
 
   def build_report
     if params[:report] == "book-popularity"
-      @books = Book.all if params[:genre].empty? || params[:genre] == 'all'
-      @books ||= Book.where(genre_id: params[:genre])
-      @books.reject!{|b| b.loans.count < 1}
-      @report_title = "Book Popularity"
-      @report_description = "These are all the books that have had any loans during the specified time period (currently since the beginning of time)."
+      set_date_range
+      build_book_popularity
+    elsif params[:report] == "unpopular-books"
+      set_date_range
+      build_unpopular_books
+    elsif params[:report] == "missing-books"
+      build_missing_books
     end
+
     if @books
       render "view_report.html.erb"
     else
-      render json: {status: :error}
+      flash[:error] = "No books match this query"
+      redirect_to :back
     end
   end
 
   def view_report
   end
 
+  private
+  def set_date_range
+    case params[:dates]
+    when "last-week"
+      @date_range = 1.week.ago.to_date..Date.today
+    when "last-month"
+      @date_range = 1.month.ago.to_date..Date.today
+    when "last-year"
+      @date_range = 1.year.ago.to_date..Date.today
+    else
+      @date_range = 100.year.ago.to_date..Date.today
```

Comment:
If you don't want to hard-code this, you could have this method accept a class as a parameter, then retrieve the created_at date of the oldest record of that class.

```
def set_date_range(klass)
  //switch
  else
    klass.first.created_at.to_date..Date.today
```